### PR TITLE
fix: avoid double module hazard 

### LINKED ISSUE #450: Refactor to avoid dual module hacks
follow up https://github.com/dai-shi/waku/pull/449#discussion_r1477724444

We should avoid this hack completely:
https://github.com/dai-shi/waku/blob/274dc34ef2516908542672213e19c78365a5e5ed/packages/waku/src/lib/handlers/handler-dev.ts#L218-L236

AFAIR, the reason we need it is because of our hack:
https://github.com/dai-shi/waku/blob/274dc34ef2516908542672213e19c78365a5e5ed/packages/waku/src/lib/renderers/rsc-renderer.ts#L26-L27

We can grab the Vite's module graph here (or I'd just pass a DEV-only resolving function).

### PR DESCRIPTION
**Pull request by Aslemammad**
Resolves #450

We warm up the vite module graph before serving the browser and the rsc tree, so at least for the initial module graph (main.tsx and its deps) we serve unified module paths (urls). 

This way, we avoid the double module hazard hack we had before.

Later, we need also to do this for each request, rather than only for the initial request, so we avoid hmr main.tsx potential bugs.

### CODE REVIEW

**Comment by Aslemammad** [comment]
not ready for review yet, just wanted to debug ci.

**Comment by dai-shi** [comment]
> not ready for review yet, just wanted to debug ci.

does draft not run the ci?

**Comment by Aslemammad** [comment]
Not at that moment. I don't know why, but I'll get it back to draft if I
don't resolve it right now.

File: packages/waku/src/lib/middleware/rsc.ts [hunk 1]
```diff
@@ -42,7 +42,10 @@ export const rsc: Middleware = (options) => {
           contentType: headers['content-type'] || '',
         };
         const readable = await (ctx.devServer
-          ? ctx.devServer.renderRscWithWorker(args)
+          ? ctx.devServer.renderRscWithWorker({
+              ...args,
+              initialModuleGraph: ctx.devServer.initialModuleGraph,
+            })
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc.ts:48]: We should make `args` for both for DEV and PRD:
```suggestion
          ? ctx.devServer.renderRscWithWorker(args, {
              initialModuleGraph: ctx.devServer.initialModuleGraph,
            })
```

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ import {
   fileURLToFilePath,
 } from '../utils/path.js';
 import { encodeInput, hasStatusCode } from './utils.js';
+import path from 'path';
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:22]:
```diff
- import path from 'path';
+ import path from 'node:path';
```
But, this file should be independent from Node.js.
That's why we made `utils/path.ts`.

File: packages/waku/src/lib/middleware/types.ts [hunk 3]
```diff
@@ -1,10 +1,13 @@
+import type { ModuleNode } from 'vite';
 import type { Config } from '../../config.js';
 import type { EntriesPrd } from '../../server.js';
 import type {
   renderRscWithWorker,
   getSsrConfigWithWorker,
 } from '../renderers/dev-worker-api.js';
 
+export type ClonableModuleNode = Pick<ModuleNode, 'url' | 'file'>;
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/types.ts:9]: Can you define type simply without Pick? Seems like the values are just strings.

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 4]
```diff
@@ -64,6 +59,8 @@ const createStreamPair = (): [Writable, Promise<ReadableStream | null>] => {
   return [writable, promise];
 };
 
+const externalizedWakuClientModules = ['waku/client', 'waku/router/client'];
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:62]: I wonder if this limits third-party libraries.
What would happen without `waku/router/client` here?

File: packages/waku/src/lib/middleware/types.ts [hunk 5]
```diff
@@ -26,6 +29,7 @@ export type HandlerContext = {
   readonly context: RscContext;
   devServer?: {
     rootDir: string;
+    initialModuleGraph: ClonableModuleNode[];
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/types.ts:32]: I wouldn't call it "Graph".

**Review by dai-shi — CHANGES_REQUESTED** [review]
Thanks for working on it. Not exactly sure how it works/solves, but I think I get the flow.

Can you merge `main` branch too?

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 4 above)
Comment (dai-shi) [reply packages/waku/src/lib/middleware/dev-server.ts:62]: This is really bad, I guess. We shouldn't depend on `waku/router` like this. Otherwise, other router integration like tanstack/router wouldn't be possible.
I guess the solution is to find a dependency from `main.tsx`.

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 6]
```diff
@@ -96,10 +94,9 @@ export const devServer: Middleware = (options) => {
       ssr: {
         external: [
           'waku',
-          'waku/client',
           'waku/server',
-          'waku/router/client',
           'waku/router/server',
+          ...externalizedWakuClientModules,
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:99]: This is for different purpose (just for out monorepo setup), so we can revert.

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 4 above)
Comment (dai-shi) [reply packages/waku/src/lib/middleware/dev-server.ts:62]: Even for `waku/client`, we know the dependency from `main.tsx`, no?

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 7]
```diff
@@ -161,13 +158,33 @@ export const devServer: Middleware = (options) => {
...
   return async (ctx, next) => {
     const [{ middleware: _removed, ...config }, vite] = await Promise.all([
       configPromise,
       vitePromise,
     ]);
+
+    const mainJs = `${config.srcDir}/${config.mainJs}`;
+    // pre-process the mainJs file to see which modules are being sent to the browser by vite
+    // and using the same modules if possible in the bundlerConfig in the stream
+    for (const m of [mainJs, ...externalizedWakuClientModules]) {
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:172]: So, start with `mainJs` and warm up all dependency modules in the graph.

File: packages/waku/src/lib/middleware/rsc.ts [hunk 8]
```diff
@@ -42,7 +42,10 @@ export const rsc: Middleware = (options) => {
           contentType: headers['content-type'] || '',
         };
         const readable = await (ctx.devServer
-          ? ctx.devServer.renderRscWithWorker(args)
+          ? ctx.devServer.renderRscWithWorker({
+              ...args,
+              initialModules: ctx.devServer.initialModules,
+            })
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc.ts:48]: Let's change the signature:
```suggestion
          ? ctx.devServer.renderRscWithWorker(args, {
              initialModules: ctx.devServer.initialModules,
            })
```

File: packages/waku/src/lib/middleware/ssr.ts [hunk 9]
```diff
@@ -61,6 +61,7 @@ export const ssr: Middleware = (options) => {
                     config,
                     pathname,
                     searchParams,
+                    initialModules: devServer.initialModules,
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/ssr.ts:64]: This too:
```suggestion
                    }, {initialModules: devServer.initialModules,
```

File: packages/waku/src/lib/renderers/dev-worker-api.ts [hunk 10]
```diff
@@ -125,7 +128,7 @@ export function registerHotUpdateCallback(
 let nextId = 1;
 
 export async function renderRscWithWorker(
-  args: RenderRscArgs,
+  args: RenderRscArgs & { initialModules: ClonableModuleNode[] },
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-api.ts:131]:
```suggestion
  args: RenderRscArgs,
  opts: { initialModules: ClonableModuleNode[] },
```

File: packages/waku/src/lib/renderers/dev-worker-api.ts [hunk 11]
```diff
@@ -182,7 +186,9 @@ export async function renderRscWithWorker(
   });
 }
 
-export async function getSsrConfigWithWorker(args: GetSsrConfigArgs): Promise<{
+export async function getSsrConfigWithWorker(
+  args: GetSsrConfigArgs & { initialModules: ClonableModuleNode[] },
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-api.ts:190]:
```suggestion
  args: GetSsrConfigArgs,
  opts: { initialModules: ClonableModuleNode[] },
```

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 12]
```diff
@@ -259,12 +259,14 @@ export const renderHtml = async (
           {},
           {
             get(_target, name: string) {
-              const file = filePath.slice(config.basePath.length);
+              const file =
+                '/' + filePath.slice(config.basePath.length).split('?')[0]!;
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:263]: it smells something.

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 13]
```diff
@@ -97,7 +99,14 @@ export async function renderRsc(
...
         const [file, name] = encodedId.split('#') as [string, string];
-        const id = resolveClientEntry(file, config);
+        let id = resolveClientEntry(file, config);
+        if (isDev) {
+          for (const moduleNode of opts.initialModules) {
+            if (moduleNode.file === file) {
+              id = moduleNode.url;
+            }
+          }
+        }
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:109]: It's better to add DEV-only code here. Can you patch `resolveClientEntry` in the caller?

**Review by dai-shi — CHANGES_REQUESTED** [review]
Please see comments.

File: packages/waku/src/lib/renderers/html-renderer.ts (hunk 12 above)
Comment (Aslemammad) [reply packages/waku/src/lib/renderers/html-renderer.ts:263]: what's that smell?

Comment (dai-shi) [reply packages/waku/src/lib/renderers/html-renderer.ts:263]: that it doesn't work on edge cases, might be the case with windows. not sure.
why do we need to add `/` prefix?

Comment (Aslemammad) [reply packages/waku/src/lib/renderers/html-renderer.ts:263]: to make it a vite path. let's remove it, even before that, it was failing, so this has no effect.

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 14]
```diff
@@ -161,13 +157,35 @@ export const devServer: Middleware = (options) => {
...
+    // and using the same modules if possible in the bundlerConfig in the stream
+    const mainJs = `${config.srcDir}/${config.mainJs}`;
+    const resolved = await vite.pluginContainer.resolveId(mainJs);
+    await vite.transformRequest(resolved!.id);
+    const resolvedModule = vite.moduleGraph.idToModuleMap.get(resolved!.id)!;
+    for (const { id } of resolvedModule.importedModules) {
+      if (id) {
+        await vite.warmupRequest(id);
+      }
+    }
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:178]: Shouldn't we use `Promise.all`?

File: packages/waku/src/lib/middleware/rsc.ts [hunk 15]
```diff
@@ -42,7 +42,14 @@ export const rsc: Middleware = (options) => {
           contentType: headers['content-type'] || '',
         };
         const readable = await (ctx.devServer
-          ? ctx.devServer.renderRscWithWorker(args)
+          ? ctx.devServer.renderRscWithWorker(
+              {
+                ...args,
+              },
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/rsc.ts:48]:
```suggestion
              args,
```

File: packages/waku/src/lib/renderers/dev-worker-impl.ts [hunk 16]
```diff
@@ -49,7 +49,13 @@ const resolveClientEntryForDev = (id: string, config: { basePath: string }) => {
 };
 
 const handleRender = async (mesg: MessageReq & { type: 'render' }) => {
-  const { id, type: _removed, hasModuleIdCallback, ...rest } = mesg;
+  const {
+    id,
+    type: _removed,
+    hasModuleIdCallback,
+    initialModules: initialModules,
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-impl.ts:56]:
```suggestion
    initialModules,
```

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 17]
```diff
@@ -259,7 +259,9 @@ export const renderHtml = async (
           {},
           {
             get(_target, name: string) {
-              const file = filePath.slice(config.basePath.length);
+              const file = filePath
+                .slice(config.basePath.length)
+                .split('?')[0]!;
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:264]: Could removing after `?` be DEV-only if it's not required for PRD?

**Review by dai-shi — COMMENTED** [review]
The diff looks good and clean. Much more readable.
Can you make a few adjustments?

File: packages/waku/src/lib/renderers/dev-worker-impl.ts [hunk 18]
```diff
@@ -73,8 +79,15 @@ const handleRender = async (mesg: MessageReq & { type: 'render' }) => {
         isDev: true,
         loadServerFile,
         loadServerModule,
-        resolveClientEntry: (id: string) =>
-          resolveClientEntryForDev(id, rest.config),
+        resolveClientEntry: (file: string) => {
+          let id = resolveClientEntryForDev(file, rest.config);
+          for (const moduleNode of initialModules) {
+            if (moduleNode.file === file) {
+              id = moduleNode.url;
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-impl.ts:86]:
```suggestion
              id = moduleNode.url;
              break;
```

Comment (dai-shi) [reply packages/waku/src/lib/renderers/dev-worker-impl.ts:86]: I think you probably have, but have you tried adding console.log here?
```suggestion
              console.log('replacing id', id, moduleNode.url);
              id = moduleNode.url;
              break;
```

Does it make any changes between ubuntu and windows?

Comment (Aslemammad) [reply packages/waku/src/lib/renderers/dev-worker-impl.ts:86]: yea tried, let me try again

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 19]
```diff
@@ -161,13 +157,35 @@ export const devServer: Middleware = (options) => {
...
+    const resolved = await vite.pluginContainer.resolveId(mainJs);
+    await vite.transformRequest(resolved!.id);
+    const resolvedModule = vite.moduleGraph.idToModuleMap.get(resolved!.id)!;
+    await Promise.all(
+      [...resolvedModule.importedModules].map(({ id }) =>
+        id ? vite.warmupRequest(id) : null,
+      ),
+    );
+
+    if (!initialModules) {
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:180]: Do we need this `if` condition? Can we or should we compute initialModules for each request? Otherwise, should we move this before L170? Or, can we even move this logic outside of this function? `vitePromise.then(...)`

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 20]
```diff
@@ -259,9 +259,11 @@ export const renderHtml = async (
           {},
           {
             get(_target, name: string) {
-              const file = filePath.slice(config.basePath.length);
+              console.log('filePath', filePath);
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:262]: Let's log `wakuDist` in L270 and see if L274 check works the same between ubuntu and windows.

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 19 above)
Comment (Aslemammad) [reply packages/waku/src/lib/middleware/dev-server.ts:180]: I remember that calling in each request gave an issue that I don't remember, but let me try it again.

File: packages/waku/src/lib/renderers/html-renderer.ts (hunk 20 above)
Comment (Aslemammad) [reply packages/waku/src/lib/renderers/html-renderer.ts:262]: sure, i'll try that soon.

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 19 above)
Comment (Aslemammad) [reply packages/waku/src/lib/middleware/dev-server.ts:180]: The whole problem lies down to the fact that in the initial load, vite has not decided what to do for modules yet, and when things happen in parallel, it counts them as two different modules even if they are the same files.

But I wonder if that's going to cause an issue for HMR in the mainJS file.

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 21]
```diff
@@ -74,6 +69,7 @@ export const devServer: Middleware = (options) => {
   const configPromise = resolveConfig(options.config);
   const vitePromise = configPromise.then(async (config) => {
     const mergedViteConfig = await mergeUserViteConfig({
+      cacheDir: 'node_modules/.vite/waku-dev-server',
```
Comment (Aslemammad) [inline packages/waku/src/lib/middleware/dev-server.ts:72]: Since we have multiple instances of vite, different ones might overwrite the others' cache. That's why we change it for this one.

File: packages/waku/src/lib/renderers/dev-worker-impl.ts [hunk 22]
```diff
@@ -73,8 +89,8 @@ const handleRender = async (mesg: MessageReq & { type: 'render' }) => {
         isDev: true,
         loadServerFile,
         loadServerModule,
-        resolveClientEntry: (id: string) =>
-          resolveClientEntryForDev(id, rest.config),
+        resolveClientEntry: (file: string) =>
+          resolveClientEntryForDev(file, rest.config, initialModules),
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-impl.ts:93]: If renaming to `file` is intentional,

File: packages/waku/src/lib/renderers/dev-worker-impl.ts [hunk 23]
```diff
@@ -42,14 +43,29 @@ const configSrcDir = getEnvironmentData('CONFIG_SRC_DIR') as string;
 const configEntriesJs = getEnvironmentData('CONFIG_ENTRIES_JS') as string;
 const configPrivateDir = getEnvironmentData('CONFIG_PRIVATE_DIR') as string;
 
-const resolveClientEntryForDev = (id: string, config: { basePath: string }) => {
+const resolveClientEntryForDev = (
+  id: string,
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/dev-worker-impl.ts:47]: this should be `file` too?

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 24]
```diff
@@ -259,12 +260,13 @@ export const renderHtml = async (
           {},
           {
             get(_target, name: string) {
-              const file = filePath.slice(config.basePath.length);
+              let file = filePath.slice(config.basePath.length);
               // TODO too long, we need to refactor this logic
               if (isDev) {
+                file = file.split('?')[0]!;
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:266]: Is this needed for both cases, a) startsWith('@fs/') and b) not?

**Review by dai-shi — COMMENTED** [review]
looking good.

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 21 above)
Comment (dai-shi) [reply packages/waku/src/lib/middleware/dev-server.ts:72]: Can you make it an inline comment in the code?

File: packages/waku/src/lib/renderers/dev-worker-impl.ts (hunk 22 above)
Comment (Aslemammad) [reply packages/waku/src/lib/renderers/dev-worker-impl.ts]: changed it for debugging, now i reverted it.

File: packages/waku/src/lib/renderers/dev-worker-impl.ts (hunk 23 above)
Comment (Aslemammad) [reply packages/waku/src/lib/renderers/dev-worker-impl.ts:47]: no need.

File: packages/waku/src/lib/renderers/html-renderer.ts (hunk 24 above)
Comment (Aslemammad) [reply packages/waku/src/lib/renderers/html-renderer.ts]: rewrote it. It's for b only.

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 21 above)
Comment (Aslemammad) [reply packages/waku/src/lib/middleware/dev-server.ts:73]: done

File: packages/waku/src/lib/middleware/dev-server.ts (hunk 19 above)
Comment (dai-shi) [reply packages/waku/src/lib/middleware/dev-server.ts:181]: Can we move this before L169?

**Review by dai-shi — APPROVED** [review]
Looks pretty good. Just one comment left.

File: packages/waku/src/lib/middleware/dev-server.ts [hunk 25]
```diff
@@ -161,13 +158,35 @@ export const devServer: Middleware = (options) => {
...
+      const resolved = await vite.pluginContainer.resolveId(mainJs);
+      const resolvedModule = vite.moduleGraph.idToModuleMap.get(resolved!.id)!;
+      await Promise.all(
+        [...resolvedModule.importedModules].map(({ id }) =>
+          id ? vite.warmupRequest(id) : null,
+        ),
+      );
+
+      initialModules = Array.from(vite.moduleGraph.idToModuleMap.values()).map(
+        (m) => ({ url: m.url, file: m.file! }),
```
Comment (dai-shi) [inline packages/waku/src/lib/middleware/dev-server.ts:183]: > `m.file!`

Is there a case when `m.file` is not defined? Just curious.

Comment (Aslemammad) [reply packages/waku/src/lib/middleware/dev-server.ts:183]: I didn't find one!

**Comment by dai-shi** [comment]
There might be some issues in some edge cases, but hope @Aslemammad can fix it soon in that case. Let's merge it.
